### DISCUSSION TITLE
External Identity Management Plugin - Creating Policies to Match Claims

### DISCUSSION THREAD

**Discussion by NotSoWell** [minio/minio#20106] [Q&A] [answered]
In the official MinIO documentation, when configuring the "External Identity Management Plugin", there is a heading "Creating Policies to Match Claims" (https://min.io/docs/minio/linux/administration/identity-access-management/pluggable-authentication.html#id5). 

At the same time, when configuring MinIO, a policy is explicitly specified that will be applied to all authenticated users (MINIO_IDENTITY_PLUGIN_ROLE_POLICY). 

I would like to know - is it possible to assign policies depending on the claim or is this a mistake in the documentation?

---
edit 17.07.2024 14:05

I already tried claims below. None of them are working.
I tested it by setting `MINIO_IDENTITY_PLUGIN_ROLE_POLICY="diagnostics"`, passing claims below from external identity manager and trying to get file.
```c#
                //{ "policy", "readonly" }, // not working
                //{ "users", "User1" }, // not working
                //{ "groups", "group1" }, // not working

                //{ "policy", "arn:minio:iam:::policy/readwrite" }, // not working
                //{ "users", "arn:minio:iam:::user/User1" }, // not working

                { "policy", "arn:aws:iam:::policy/readwrite" }, // not working
                { "users", "arn:aws:iam:::user/User1" } // not working
```

**Comment by dridk** [comment]
Did you solve it ?  I have no idea what to use in "claims" fields .. It says KEY=VALUE  .. What is the key what is the value ?

**Reply by dridk** [reply]
https://github.com/minio/minio/blob/d7ef6315aea1a72f0c55828b6d6a19f915382647/docs/iam/identity-manager-plugin.go#L49

**Reply by dridk** [reply]
Similar issue than https://github.com/minio/minio/discussions/20369

**Reply by NotSoWell** [reply, marked as answer]
I think there is an error in documentation. You can't dynamically set policy by matching claims. When using "MinIO External Identity Management Plugin" you set global and **only one policy for all external clients** in "MINIO_IDENTITY_PLUGIN_ROLE_POLICY" variable.

About claims - you can use them to make one flexible policy. Read this article "https://min.io/docs/minio/linux/administration/identity-access-management/policy-based-access-control.html#policy-variables". You can use claims like "OpenID Policy Variables". Example below from minio docs. By passing in claims `"preferred_username": "john"` this variable `${jwt:preferred_username}` in policy will be replaced with value "john"
```json
{
"Version": "2012-10-17",
"Statement": [
      {
         "Action": ["s3:ListBucket"],
         "Effect": "Allow",
         "Resource": ["arn:aws:s3:::mybucket"],
         "Condition": {"StringLike": {"s3:prefix": ["${jwt:preferred_username}/*"]}}
      },
      {
         "Action": [
         "s3:GetObject",
         "s3:PutObject"
         ],
         "Effect": "Allow",
         "Resource": ["arn:aws:s3:::mybucket/${jwt:preferred_username}/*"]
      }
   ]
}
```
